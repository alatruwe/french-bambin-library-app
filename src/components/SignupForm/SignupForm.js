import React, { Component } from "react";
import { withRouter } from "react-router-dom";

import ValidationError from "../ValidationError/ValidationError.js";

class SignupForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      firstName: {
        value: "",
        touched: false,
      },
      lastName: {
        value: "",
        touched: false,
      },
      email: {
        value: "",
        touched: false,
      },
      password: {
        value: "",
        touched: false,
      },
      repeatPassword: {
        value: "",
        touched: false,
      },
    };
  }

  updateFirstName(name) {
    this.setState({ firstName: { value: name, touched: true } });
  }
  updateLastName(name) {
    this.setState({ lastName: { value: name, touched: true } });
  }
  updateEmail(email) {
    this.setState({ email: { value: email, touched: true } });
  }
  updatePassword(password) {
    this.setState({
      password: {
        value: password,
        touched: true,
      },
    });
  }
  updateRepeatPassword(repeatPassword) {
    this.setState({
      repeatPassword: { value: repeatPassword, touched: true },
    });
  }

  validateFirstName() {
    const firstName = this.state.firstName.value.trim();
    if (firstName.length === 0) {
      return "Name is required";
    } else if (firstName.length < 3) {
      return "Name must be at least 3 characters long";
    }
  }
  validateLastName() {
    const lastName = this.state.lastName.value.trim();
    if (lastName.length === 0) {
      return "Name is required";
    } else if (lastName.length < 3) {
      return "Name must be at least 3 characters long";
    }
  }
  validateEmail() {
    const email = this.state.email.value.trim();
    if (email.length === 0) {
      return "Email is required";
    }
  }
  validatePassword() {
    const password = this.state.password.value.trim();
    if (password.length === 0) {
      return "Password is required";
    } else if (password.length < 6 || password.length > 72) {
      return "Password must be between 6 and 72 characters long";
    } else if (!password.match(/[0-9]/)) {
      return "Password must contain at least one number";
    }
  }
  validateRepeatPassword() {
    const repeatPassword = this.state.repeatPassword.value.trim();
    const password = this.state.password.value.trim();

    if (repeatPassword !== password) {
      return "Passwords do not match";
    }
  }

  handleSubmit(event) {
    event.preventDefault();
    this.props.history.push(`/home`);
    this.props.handleAuthSubmit();
  }

  render() {
    const firstNameError = this.validateFirstName();
    const lastNameError = this.validateLastName();
    const emailError = this.validateEmail();
    const passwordError = this.validatePassword();
    const repeatPasswordError = this.validateRepeatPassword();

    return (
      <form onSubmit={(e) => this.handleSubmit(e)}>
        <div className="form-group">
          <label htmlFor="name">First name :</label>
          <input
            type="text"
            name="firstName"
            id="firstName"
            onChange={(e) => this.updateFirstName(e.target.value)}
          />
          {this.state.firstName.touched && (
            <ValidationError message={firstNameError} />
          )}
        </div>
        <div className="form-group">
          <label htmlFor="name">Last name :</label>
          <input
            type="text"
            name="lastName"
            id="lastName"
            onChange={(e) => this.updateLastName(e.target.value)}
          />
          {this.state.lastName.touched && (
            <ValidationError message={lastNameError} />
          )}
        </div>
        <div className="form-group">
          <label htmlFor="name">Email :</label>
          <input
            type="email"
            name="email"
            id="email"
            onChange={(e) => this.updateEmail(e.target.value)}
          />
          {this.state.email.touched && <ValidationError message={emailError} />}
        </div>
        <div className="form-group">
          <label htmlFor="password">Password *</label>
          <input
            type="password"
            name="password"
            id="password"
            onChange={(e) => this.updatePassword(e.target.value)}
          />

          {this.state.password.touched && (
            <ValidationError message={passwordError} />
          )}
        </div>
        <div className="form-group">
          <label htmlFor="repeatPassword">Repeat Password *</label>
          <input
            type="password"
            name="repeatPassword"
            id="repeatPassword"
            onChange={(e) => this.updateRepeatPassword(e.target.value)}
          />
          {this.state.repeatPassword.touched && (
            <ValidationError message={repeatPasswordError} />
          )}
        </div>

        <div className="registration__button__group">
          <button
            type="submit"
            disabled={
              this.validateFirstName() ||
              this.validateLastName() ||
              this.validateEmail() ||
              this.validatePassword() ||
              this.validateRepeatPassword()
            }
          >
            Sign Up!
          </button>
        </div>
      </form>
    );
  }
}

export default withRouter(SignupForm);
