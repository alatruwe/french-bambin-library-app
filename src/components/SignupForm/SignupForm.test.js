import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import SignupForm from "./SignupForm";

describe(`SignupForm Component`, () => {
  it("renders without errors", () => {
    const div = document.createElement("div");
    ReactDOM.render(<SignupForm />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders the UI as expected", () => {
    const tree = renderer.create(<SignupForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
