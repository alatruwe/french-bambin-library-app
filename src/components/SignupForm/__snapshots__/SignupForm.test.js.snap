// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm Component renders the UI as expected 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      htmlFor="name"
    >
      First name :
    </label>
    <input
      id="firstName"
      name="firstName"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="name"
    >
      Last name :
    </label>
    <input
      id="lastName"
      name="lastName"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="name"
    >
      Email :
    </label>
    <input
      id="email"
      name="email"
      onChange={[Function]}
      type="email"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="password"
    >
      Password *
    </label>
    <input
      id="password"
      name="password"
      onChange={[Function]}
      type="password"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="repeatPassword"
    >
      Repeat Password *
    </label>
    <input
      id="repeatPassword"
      name="repeatPassword"
      onChange={[Function]}
      type="password"
    />
  </div>
  <div
    className="registration__button__group"
  >
    <button
      disabled="Name is required"
      type="submit"
    >
      Sign Up!
    </button>
  </div>
</form>
`;
